File: 28-12-2024_Find_all_Triplets_with_Zero_Sum.java
class Solution {
    public List<List<Integer>> findTriplets(int[] arr) {
        Set<List<Integer>> resultSet = new HashSet<>();
    
    int len = arr.length;
    
    Map<Integer, List<int[]>> sumWithPair = new HashMap<>();
    
    for (int i = 0; i < len; i++) {
        for (int j = i + 1; j < len; j++) {
            int sum = arr[i] + arr[j];
            
            sumWithPair.putIfAbsent(sum, new ArrayList<>());
            sumWithPair.get(sum).add(new int[]{i, j});
        }
    }
    
    for (int i = 0; i < len; i++) {
        int complement = -arr[i];
        
        if (sumWithPair.containsKey(complement)) {
            for (int[] pair : sumWithPair.get(complement)) {
                if (i != pair[0] && i != pair[1]) {
                    List<Integer> triplet = new ArrayList<>(Arrays.asList(i, pair[0], pair[1]));
                    Collections.sort(triplet);
                    resultSet.add(triplet);
                }
            }
        }
    }
    
    return new ArrayList<>(resultSet);
    }
}

File: 28-12-2024_Find_All_Triplets_with_Zero_Sum.py
class Solution:
    def findTriplets(self, arr):
        result_set = set()
        
        n = len(arr)
        
        sum_with_pair = {}
        
        for i in range(n):
            for j in range(i + 1, n):
                total = arr[i] + arr[j]
                
                if total not in sum_with_pair:
                    sum_with_pair[total] = []
                sum_with_pair[total].append((i, j))
        
        for i in range(n):
            complement = -arr[i]
            
            if complement in sum_with_pair:
                for pair in sum_with_pair[complement]:
                    if i != pair[0] and i != pair[1]:
                        triplet = sorted([i, pair[0], pair[1]])
                        result_set.add(tuple(triplet))
        
        return [list(triplet) for triplet in result_set]
